idaes/apps/uncertainty_propagation/sens.py

```diff
@@ -0,0 +1,723 @@
+            var_dic[i] = i
+    for v in theta_names:
+        v_tmp = str(kk)
+        original_param_object = Param(initialize=theta[v], mutable=True)
+        perturbed_param_object = Param(initialize=theta[v])
+        m.add_component("original_"+v_tmp, original_param_object)
+        m.add_component("perturbed_"+v_tmp, perturbed_param_object)
+        m.extra.add(eval('m.'+var_dic[v]) - eval('m.original_'+v_tmp) == 0 )
+        original_Param.append(original_param_object)
+        perturbed_Param.append(perturbed_param_object)
+        kk = kk + 1
+    m_kaug_dsdp = sensitivity_calculation('kaug',m,original_Param,perturbed_Param, tee)
+
+    try:
+        with open ("./dsdp/col_row.col", "r") as myfile:
+            col = myfile.read().splitlines()
+        with open ("./dsdp/col_row.row", "r") as myfile:
+            row = myfile.read().splitlines()
+        dsdp = np.loadtxt("./dsdp/dsdp_in_.in")
+    except Exception as e:
+        print('File not found.')
+    dsdp = dsdp.reshape((len(theta_names), int(len(dsdp)/len(theta_names))))
+    dsdp = dsdp[:len(theta_names), :len(col)]
+    try:
+        shutil.rmtree('dsdp', ignore_errors=True)
+    except OSError:
+        pass
+    col = [i for i in col if SensitivityInterface.get_default_block_name() not in i]
+    return dsdp, col
+
+def get_dfds_dcds(model, theta_names, tee=False, solver_options=None):
+    """This function calculates gradient vector of the objective function 
+       and constraints with respect to the variables in theta_names.
+       We assume that the number of  constraints (rows) = Ncon, 
+       the number of theta = Ntheta, the numer of decision (primal) variables = Nx. 
+       Then, the number of all variables (columns) = Nvar = Ntheta + Nx. 
+       All results follow row and column order in row and col.
+
+    e.g) min f:  p1*x1+ p2*(x2^2) + p1*p2
+         s.t  c1: x1 + x2 = p1
+              c2: x2 + x3 = p2
+              0 <= x1, x2, x3 <= 10
+              p1 = 10
+              p2 = 5
+    - Variables = (x1, x2, x3, p1, p2)
+    - Fix p1 and p2 with estimated values
+
+    Parameters
+    ----------
+    model: Pyomo ConcreteModel
+        model should includes an objective function 
+    theta_names: list of strings
+        List of Var names
+    tee: bool, optional
+        Indicates that ef solver output should be teed
+    solver_options: dict, optional
+        Provides options to the solver (also the name of an attribute)
+    
+    Returns
+    -------
+    gradient_f: numpy.ndarray
+        Length Nvar array. A gradient vector of the objective function with respect to the (decision variables, parameters) at the optimal solution
+    gradient_c: numpy.ndarray 
+        Ncon by Nvar size matrix. A Jacobian matrix of the constraints with respect to the (decision variables, parameters) at the optimal solution
+        Each row contains [column number, row number, and value], colum order follows variable order in col and index starts from 1. 
```

Comment:
The index should start at 0. Otherwise, the numpy array has the wrong dimensions. My understanding is that `gradient_f` and `gradient_c` are sparse numpy arrays. If you start the index at 1, they will have the wrong dimensions.

Reply:
- I think the numpy array is better for `gradient_f`.
- I agree with you for `gradient_c`. I updated it to the scipy sparse form.

Reply:
I also changed `dsdp` to scipy sparse form